src/WebJobs.Script.WebHost/Controllers/KeysController.cs

```diff
@@ -249,12 +249,18 @@ private async Task<IActionResult> AddOrUpdateSecretAsync(string keyName, string
 
         private async Task<IActionResult> DeleteFunctionSecretAsync(string keyName, string keyScope, ScriptSecretsType secretsType)
         {
-            if (keyName == null || keyName.StartsWith("_"))
+            if (keyName == null)
             {
-                // System keys cannot be deleted.
                 return BadRequest("Invalid key name.");
             }
 
+            HostSecretsInfo hostSecrets = await _secretManagerProvider.Current.GetHostSecretsAsync();
+            if (hostSecrets.SystemKeys != null && hostSecrets.SystemKeys.TryGetValue(keyName, out string _))
```

Comment:
SystemKeys is an ARM API modifiable collection, so we can't prevent deletion like this. SystemKeys contain more than just the _master key. E.g. they're used by custom webhook extensions. By convention, currently those keys won't currently start with an underscore (code [here](https://github.com/Azure/azure-functions-host/blob/3bec05a840f18989b37b998d7cf904f0fcdca848/src/WebJobs.Script.WebHost/WebHooks/DefaultScriptWebHookProvider.cs#L90)).

Also, it doesn't look like we've locked down the create path for keys to officially prevent underscores. E.g. you can create both system and regular keys yourself via ARM APIs prefixed with underscores. Hence this issue.

Since we're just trying to prevent deletion of the _master key, I wonder if it wouldn't be safer for us to just check against a list of disallowed system keys here, e.g. have a function IsBuiltInSystemKeyName initially containing only "_master".

Reply:
Thanks. Makes sense. Updated PR.!